core/src/solution_submission.rs

```diff
@@ -142,40 +146,27 @@ impl<'a> StableXSolutionSubmitter<'a> {
         SolutionSubmissionError::Unexpected(err.into())
     }
 
-    async fn handle_submit_solution_result(
+    async fn cancel_transaction_after_deadline(
         &self,
         batch_index: u32,
-        solution: Solution,
-        result: std::result::Result<Result<(), MethodError>, TimeoutError>,
         nonce: U256,
-    ) -> Result<(), SolutionSubmissionError> {
-        if let Ok(submit_result) = result {
-            match submit_result {
-                Ok(()) => Ok(()),
-                Err(err) => Err(self.make_error(batch_index, solution, err).await),
-            }
-        } else {
-            let gas_price =
-                U256::from((GAS_PRICE_CAP as f64 * MIN_GAS_PRICE_INCREASE_FACTOR).ceil() as u128);
-            log::info!(
-                "cancelling transaction because it took too long, using gas price {}",
-                gas_price
-            );
-            match self.contract.send_noop_transaction(gas_price, nonce).await {
-                Ok(_) => log::info!(
-                    "cancelled solution submission of batch {} because of deadline",
-                    batch_index
-                ),
-                Err(err) => log::error!(
-                    "failed to cancel solution submission of batch {} after deadline: {:?}",
-                    batch_index,
-                    err
-                ),
-            }
-            Err(SolutionSubmissionError::Unexpected(anyhow!(
-                "solution submission transaction not confirmed in time"
-            )))
-        }
+    ) -> Result<(), NoopTransactionError> {
+        // Add some extra time in case of desync between real time and ethereum node current block time.
+        let deadline = BatchId::from(batch_index).solve_end_time() + Duration::from_secs(30);
+        let remaining = deadline
+            .duration_since(SystemTime::now())
+            .unwrap_or(Duration::from_secs(0));
+        self.async_sleep.sleep(remaining).await;
```

Comment:
Should we still attempt the cancellation even if the other future has already completed? Or should we have some conditional variable that we signal when the tx gets confirmed to avoid causing unnecessary failure on the node?

Reply:
The way this works is that both futures are created and then "race" in the call to `select`. But the cancel future only sends the noop transaction after sleeping until the deadline. So if the solution submission future completes first the cancel futures is dropped and will never send the noop transaction.

Reply:
Yup! The sleep + cancel future will stop being `poll`-ed once the submit future completes and therefore stop progressing.

There is a subtle case where the cancel transaction will just have been sent when the submit solutions transaction completes. However I believe this case will also be gracefully handled as the node will deny the transaction (since it had the same nonce as a mined tx) and the deadline future won't produce an error as it will not be `poll`-ed further, causing it not to handle the nodes error response.

Reply:
That makes sense and sounds good. I was just worried that we will always send a noop tx on every submission which doesn't seem to be the case. I don't know if the "drop" behavior can be easily tested now that we control the sleep. If so it would be nice to add, but otherwise never mind.